torch/csrc/autograd/engine.h

```diff
@@ -68,6 +71,22 @@ struct TORCH_API Engine {
   std::vector<std::function<void()>> final_callbacks_;
   // To protect reads and writes to final_callbacks_
   std::mutex post_callbacks_lock_;
+
+  struct ThreadPoolShared {
+    std::atomic_int num_workers_;
+    std::condition_variable work_;
+    std::mutex graphtasks_queue_lock_;
+    std::queue<std::shared_ptr<GraphTask>> graphtasks_queue_;
+
+    ThreadPoolShared() : num_workers_(0) {};
+ };
+
+ // Temporary workaround until shutting down threads is done
+ // Need shared ownership of all these objects because threads are leaked
+ // when Engine shuts down
+ std::shared_ptr<ThreadPoolShared> thread_pool_shared_;
+ // TODO don't really need this because the parent thread waits
+ std::vector<std::shared_ptr<std::mutex>> device_locks_;
```

Comment:
I assume you wrapped this in a `shared_ptr` because otherwise mutex is not movable. If you can allocate `device_locks_` all at once and then never use any possibly resizing functions, the indirection is not necessary. (If you still need to indirect, `unique_ptr` is a better choice than `shared_ptr` because you don't really need shared ownership here.)

Reply:
Update: This is actually fixing the "Temporary workaround until shutting down threads is done" problem.

Reply:
I think I agree with you that the lock is not needed.